File: analysis/model/src/test/java/de/maibornwolff/codecharta/model/TreeTest.java
/*
 * Copyright (c) 2017, MaibornWolff GmbH
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  * Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of  nor the names of its contributors may be used to
 *    endorse or promote products derived from this software without specific
 *    prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package de.maibornwolff.codecharta.model;

import org.junit.Test;

import java.util.Arrays;
import java.util.Map;

import static de.maibornwolff.codecharta.model.TreeCreator.createTree;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class TreeTest {
    @Test
    public void getNodeBy_trivial_should_return_same_tree() {
        Tree tree = createTree(new Path(Arrays.asList("bla")), createTree());

        assertThat(tree.getNodeBy(Path.TRIVIAL).get(), is(tree));
    }

    @Test
    public void getNode_of_simple_tree_should_return_this_tree() {
        Tree tree = createTree();

        Map nodes = tree.getNodes();

        assertThat(nodes.size(), is(1));
        assertThat(nodes.keySet().contains(Path.TRIVIAL), is(true));
        assertThat(nodes.values().contains(tree), is(true));
    }

    @Test
    public void getLeaves_of_simple_tree_should_return_this_leaf() {
        Tree tree = createTree();

        Map leaves = tree.getLeaves();

        assertThat(leaves.size(), is(1));
        assertThat(leaves.keySet().contains(Path.TRIVIAL), is(true));
        assertThat(leaves.values().contains(tree), is(true));
    }

    @Test
    public void getNode_should_return_nodes() {
        Tree innerTree = createTree();
        Path pathToInnerTree = new Path(Arrays.asList("bla"));
        Tree tree = createTree(pathToInnerTree, innerTree);

        Map nodes = tree.getNodes();

        assertThat(nodes.size(), is(2));
        assertThat(nodes.keySet().contains(Path.TRIVIAL), is(true));
        assertThat(nodes.values().contains(tree), is(true));
        assertThat(nodes.keySet().contains(pathToInnerTree), is(true));
        assertThat(nodes.values().contains(innerTree), is(true));
    }

    @Test
    public void getLeaves_should_return_leaves() {
        Tree innerTree = createTree();
        Path pathToInnerTree = new Path(Arrays.asList("bla"));
        Tree tree = createTree(pathToInnerTree, innerTree);

        Map leaves = tree.getLeaves();

        assertThat(leaves.size(), is(1));
        assertThat(leaves.keySet().contains(pathToInnerTree), is(true));
        assertThat(leaves.values().contains(innerTree), is(true));
    }
}

File: analysis/model/src/test/kotlin/de/maibornwolff/codecharta/model/TreeTest.kt
package de.maibornwolff.codecharta.model

import de.maibornwolff.codecharta.model.TreeCreator.createTree
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import kotlin.test.assertFailsWith

@DisplayName("Tree Test")
class TreeTest {
@Nested
    @DisplayName("Tree Test > a tree of depth 0")
    inner class ZeroDepthTree {
    val tree = createTree()

        @Test
        fun `nodes should be only the tree itself`() {
            val nodes = tree.nodes
            assertThat(nodes.size).isEqualTo(1)
            assertThat(nodes.keys).contains(Path.TRIVIAL)
            assertThat(tree.asTreeNode()).isIn(nodes.values)
        }

        @Test
        fun `leaves should be only the tree itself`() {
            val leaves = tree.leaves
            assertThat(leaves.size).isEqualTo(1)
            assertThat(leaves.keys).contains(Path.TRIVIAL)
            assertThat(tree.asTreeNode()).isIn(leaves.values)
        }

        @Test
        fun `getNodeBy should return same tree on trivial path`() {
            assertThat(tree.getNodeBy(Path.TRIVIAL)).isEqualTo(tree)
        }

        @Test
        fun `getNodeBy should throw exception on path not contained in tree`() {
            assertFailsWith(NoSuchElementException::class) {
                tree.getNodeBy(Path("nonexistingpath"))
            }
        }
    }

    @Test
    fun `merging should do nothing by default`() {
        val tree = createTree()
        tree.merge(listOf(createTree().asTreeNode()))
        val leaves = tree.leaves

        assertThat(leaves).hasSize(1)
        assertThat(leaves.keys).contains(Path.TRIVIAL)
        assertThat(tree.asTreeNode()).isIn(leaves.values)
    }

    @Nested
    @DisplayName("Tree Test > a tree of depth 1")
    inner class OneDepthTree {
    val innerTree = createTree()
        val pathToInnerTree = Path("bla")
        val tree = createTree(pathToInnerTree, innerTree)

        @Test
        fun `nodes should contain exactly itself and the subtree`() {
            val nodes = tree.nodes
            assertThat(nodes).hasSize(2)
            assertThat(nodes.keys).contains(Path.TRIVIAL)
            assertThat(nodes.keys).contains(pathToInnerTree)
            assertThat(tree.asTreeNode()).isIn(nodes.values)
            assertThat(innerTree.asTreeNode()).isIn(nodes.values)
        }

        @Test
        fun `leaves should return the inner tree`() {
            val leaves = tree.leaves

            assertThat(leaves).hasSize(1)
            assertThat(leaves.keys).contains(pathToInnerTree)
            assertThat(innerTree.asTreeNode()).isIn(leaves.values)
        }

        @Test
        fun `getNodeBy should return same tree on trivial path`() {
            assertThat(tree.getNodeBy(Path.TRIVIAL)).isEqualTo(tree)
        }

        @Test
        fun `getNodeBy should return inner tree on pathToInnerTree`() {
            assertThat(tree.getNodeBy(pathToInnerTree)).isEqualTo(innerTree)
        }

        @Test
        fun `getNodeBy should throw exception on path not contained in tree`() {
            assertFailsWith(NoSuchElementException::class) {
                tree.getNodeBy(Path("nonexistingpath"))
            }
        }
    }
}
